### PR TITLE
Cable diagnostics attribute added

### PR DESCRIPTION
**Pull request by prafull-brcm**
Cable diagnostics is required when the characteristics of the problems associated with a cable become of paramount concern when a link cannot be established or is performing poorly. 
Enhanced Cable Diagnostics (ECD) provides with intelligent pre-processing and post-processing algorithm inside HW that can be initiated at any time by setting specific register bits and then reading the results.
Usage of the attribute:
sai_attr_set[0].id = SAI_PORT_ATTR_CABLE_TYPE;
sai_attr_set[0].value.s32 = SAI_PORT_CABLE_TYPE_CAT6;
ret = create_port(&port_id_0, phy_2_switch_id, 2, sai_attr_set);
or Cable type can be  set latter also using set port attribute function if port is already created.
sai_attr_set.id = SAI_PORT_ATTR_CABLE_TYPE;
sai_attr_set.value.s32 = SAI_PORT_CABLE_TYPE_CAT6;
set_port_attribute(port_id_0, &sai_attr_set);

Cable pair status and cable length is read only attribute, can be called anytime.
sai_attr_get.id = SAI_PORT_ATTR_CABLE_PAIR_STATE;
ret = get_port_attribute(port_id_0,1, &sai_attr_get);
Note: Please note that cable diagnostics is debug feature it puts chip in debug mode user needs to re-configure the mode and auto-negotiation after cable pair status and cable length attribute is requested to work in normal mode.

### CODE REVIEW

**Comment by kcudnik** [comment]
please fix build errors

**Comment by prafull-brcm** [comment]
> please fix build errors

Fixed

**Comment by rlhui** [comment]
please add description of this PR with motivation, thanks.

**Review by kcudnik — APPROVED** [review]

**Comment by lguohan** [comment]
@jiahua-wang , @prafull-brcm , 
1. please change the doc to md format and incoporate this into the PR.
2. can you clarify the example, make the create_port optional. as my understand, we can also set the port cable type to CAT6.
3. as my understanding, the diagnostic process happened during the query for the attributes. when you query status, it runs on diagnose once, when you try to get length, it will run another time.

**Comment by prafull-brcm** [comment]
> @jiahua-wang , @prafull-brcm ,
> 
> 1. please change the doc to md format and incoporate this into the PR.
> 2. can you clarify the example, make the create_port optional. as my understand, we can also set the port cable type to CAT6.
> 3. as my understanding, the diagnostic process happened during the query for the attributes. when you query status, it runs on diagnose once, when you try to get length, it will run another time.
@lguohan 
1. I incorporated the the document in the PR in md format
2. I mentioned in example that using set property cable type can be set.
3. Yes it is correct, Run diagnostics command runs for both cable pair status and cable length.
    I thought of option to add new attribute as run diag command and store the result in database and read latter using cable pair status and cable length. It will be two step process once run the command and read the result in subsequent call. There is chance of reading old result without calling run diag command this is the reason I added to run the command each time status or cable length is requested.
Let me know I am fine to either approach.

**Comment by prafull-brcm** [comment]
@lguohan 
Updated the documentation in proper md format.

File: doc/cable_diag/cable_diag.md [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+## 1 Overview
+
+When a link is up and operating normally with adequate bit error
+performance, there is no need for any diagnostic either on the
+digital/analog portion of the link or on the cable connecting the link.
+However, the characteristics of the problems associated with a cable
+become of paramount concern when a link cannot be established or is
+performing poorly.
+
+Enhanced Cable Diagnostics (ECD) provides with intelligent
```
Comment (prgeor) [inline doc/cable_diag/cable_diag.md:71]: @prafull-brcm where is ECD implemented on the hardware? On the ASIC? Serdes core?

Comment (prafull-brcm) [reply doc/cable_diag/cable_diag.md:71]: @prgeor  ECD is implemented inside the PHY chip HW.

File: inc/saiport.h [hunk 2]
```diff
@@ -166,6 +166,53 @@ typedef enum _sai_port_media_type_t
...
+
+    /**  Cable state unknown */
+    SAI_PORT_CABLE_PAIR_STATE_UNKNOWN
+} sai_port_cable_pair_state_t;
+
+/**
+ * @brief Attribute data for #SAI_PORT_ATTR_CABLE_TYPE
+ * Copper cable types
+ */
+typedef enum _sai_port_cable_type_t
```
[the remark this answers is not in the document]
Comment (prafull-brcm) [reply inc/saiport.h:195]: @prgeor Cable type information need to be provided by the user, it is not automatically detected. If no information is provided it takes cable type as Unknown.

File: inc/saiport.h [hunk 3]
```diff
@@ -2400,6 +2447,40 @@ typedef enum _sai_port_attr_t
...
+     * @brief Get ethernet copper cable pair length (unit meter)
+     *
+     * Returns ethernet cable pair length sequentially from list index 0 to n (n = number of pairs - 1)
+     * value of n depends on number of pairs in twisted ethernet copper cable
+     * Cable diagnostics runs once cable pair length is requested, it is synchronous call.
+     *
+     * @type sai_s32_list_t
+     * @flags READ_ONLY
+     */
+    SAI_PORT_ATTR_CABLE_PAIR_LENGTH,
```
Comment (prgeor) [inline inc/saiport.h:2472]: @prafull-brcm is it the length of the DAC/copper cable? If not, what exactly is this length indicate?

Comment (prafull-brcm) [reply inc/saiport.h:2472]: @prgeor It indicates copper cable length for example twisted wire pairs, coaxial cables with tolerance of +-1meter

**Review by prgeor — APPROVED** [review]

**Pull request merged**